staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -231,40 +272,59 @@ func (c *celAdmissionController) Validate(
 			// be a bindingInfo for it
 			bindingInfo := c.bindingInfos[namespacedBindingName]
 			binding := bindingInfo.lastReconciledValue
-			if !binding.Matches(a) {
+			matches, err := c.validatorCompiler.BindingMatches(a, o, binding)
+			if err != nil {
+				// Configuration error.
+				addConfigError(err, definition, binding)
+				continue
+			}
+			if !matches {
 				continue
 			}
 
 			var param *unstructured.Unstructured
 
-			// If definition has no paramsource, always provide nil params to
-			// evaluator. If binding specifies a params to use they are ignored.
-			// Done this way so you can configure params before definition is ready.
-			if paramSource := definition.GetParamSource(); paramSource != nil {
-				paramsNamespace, paramsName := binding.GetTargetParams()
+			// If definition has paramKind, paramRef is required in binding.
+			// If definition has no paramKind, paramRef set in binding will be ignored.
+			paramKind := definition.Spec.ParamKind
+			paramRef := binding.Spec.ParamRef
+			if paramKind != nil && paramRef != nil {
 
 				// Find the params referred by the binding by looking its name up
 				// in our informer for its CRD
-				paramInfo, ok := c.paramsCRDControllers[*paramSource]
+				paramInfo, ok := c.paramsCRDControllers[*paramKind]
 				if !ok {
-					addConfigError(fmt.Errorf("paramSource kind `%v` not known",
-						paramSource.String()), definition, binding)
+					addConfigError(fmt.Errorf("paramKind kind `%v` not known",
+						paramKind.String()), definition, binding)
 					continue
 				}
 
-				if len(paramsNamespace) == 0 {
-					param, err = paramInfo.controller.Informer().Get(paramsName)
+				// If the param informer for this admission policy has not yet
+				// had time to perform an initial listing, don't attempt to use
+				// it.
+				if !cache.WaitForNamedCacheSync("", ctx.Done(), paramInfo.controller.HasSynced) {
+					addConfigError(fmt.Errorf("paramKind kind `%v` not yet synced to use for admission",
+						paramKind.String()), definition, binding)
+					continue
+				}
```

Comment:
WaitForNamedCacheSync can block up until the request timeout, which could be *long*. We should not call that in the admission Validate path.

There are a few reasons the informer might not have synced yet:
1. the restmapping was not resolveable and the informer is looping on retry (this should be an error)
2. the restmapping was resolveable but there are a lot of objects and the informer has not completed syncing yet  (this should *maybe* try to fetch the param object directly once, though I'm not sure)
3. the restmapping was resolveable and the param informer *just* got created and hasn't completed the first list year (this should *maybe* try to fetch the param object directly once, though I'm not sure)

For now, check for an unsynced param informer directly and fail fast, but add a follow-up to revisit this for possible single-shot live lookup
```go
if !paramInfo.controller.HasSynced() {
		addConfigError(fmt.Errorf("paramKind kind `%v` not yet synced to use for admission",
			paramKind.String()), definition, binding)
		continue
}
```